File: nodejs/cacheitemversioning/src/cache-item-versioning.js
const {CacheManager,CacheItem,JsonDataType} = require('ncache-client');
let cache = null;
const config = require('./app.config.json');
async function Run() {
    // Initialize cache 
    await InitializeCache();
    await cache.clear();
    // Create a simple customer object
    let customer = CreateNewCustomer();
    const key = GetKey(customer);

    // Adding item
    const itemVersion = await AddObjectToCache(key, customer);

    // Modify the object and update in cache
    const latestItemVersion = await UpdateObjectInCache(key, customer,itemVersion);

    // Get an item with cache item version
    await GetObjectWithItemVersion(key,itemVersion,latestItemVersion);

    // Remove an item if cache item 
    await RemoveObjectFromCache(key,latestItemVersion);

    //close cache
    await cache.close();

}

async function InitializeCache() {
    // Get Cache Name
    const cacheName = config.cacheProperties.cacheName
    if (!cacheName) {
        console.log("The CacheID cannot be null or empty.");
        return;
    }
    // Initialize an instance of the cache to begin performing operations:
    cache = await CacheManager.getCache(cacheName);

    // Print output on console
    console.log(`Cache ${cacheName} is initialized.`);
            
}

async function AddObjectToCache(key, customer) {
    //Populating cache item
    const item = new CacheItem(customer,"Customer");

    let cacheItemVersion = await cache.add(key, item);

    // Print output on console
    console.log("\nObject is added to cache.");
    return  cacheItemVersion;
}

async function GetObjectWithItemVersion(key,oldCacheItemVersion,latestitemVersion) {
    // Retrives item from cache based on CacheItemVersion
    // Get item only is version superior to old version
    let cachedCustomer = await cache.getIfNewer(key,JsonDataType.Object,oldCacheItemVersion);
    if (cachedCustomer === null) {
        // Print output on console
        console.log("Specified item version is latest.");
    } else {
      // Print output on console
      console.log("\nObject is fetched from cache");
      PrintCustomerDetails(cachedCustomer);   
    }

    // Get an item from cache by latest item version
    cachedCustomer = await cache.getIfNewer(key,JsonDataType.Object,latestitemVersion);
    if (cachedCustomer === null) {
        // Print output on console
        console.log("Specified item version is latest.");
    } else {
      // Print output on console
      console.log("\nObject is fetched from cache");
      PrintCustomerDetails(cachedCustomer);   
    }
}

async function UpdateObjectInCache(key, customer,oldItemVersion) {
     // Updaing the customer object in cache;
     customer.CompanyName = "Gourmet Lanchonetes";
     // Create a CacheItem instance and assign Cache Item Version
     let item = new CacheItem(customer,"Customer");
     item.setCacheItemVersion(oldItemVersion);

     // Update item if version match
     const updatedItemVersion = await cache.insert(key, item);
     if (oldItemVersion.getVersion() != updatedItemVersion.getVersion()) {
        console.log("\nItem has changed since last time it was fetched.");
     }
     return updatedItemVersion;
}

async function RemoveObjectFromCache(key) {
   // Remove object from cache if cache item version matched in cache
    await cache.remove(key,JsonDataType.Object);

    // Print output on console
    console.log("\nObject is removed from cache.");
}

function GetKey(customer) {
    return `Customer:${customer.CustomerID}`;
}
function CreateNewCustomer() {
    return({
        CustomerID : "DAVJO",
        ContactName : "David Johnes",
        CompanyName : "Lonesome Pine Restaurant",
        ContactNo : "12345-6789",
        Address : "Silicon Valley, Santa Clara, California",
    });
}
function PrintCustomerDetails(customer) {
    if (customer == null) return;

        console.log();
        console.log("Customer Details are as follows: ");
        console.log("ContactName: " + customer.ContactName);
        console.log("CompanyName: " + customer.CompanyName);
        console.log("Contact No: " + customer.ContactNo);
        console.log("Address: " + customer.Address);
        console.log();

}
module.exports = {
    Run
}

File: python/cache_item_versioning/cache_item_versioning.py
from ncache.client.CacheManager import CacheManager
from ncache.client.CacheItem import CacheItem
from sample_data.product import Product


def initialize_cache():
    cache_name = "demoCache"
    cache = CacheManager.get_cache(cache_name)
    print("\nCache initialized successfully.")
    return cache


def create_product():
    product = Product(id="1001", name="Tea", unit_price=50, units_available=20)
    return product


def get_key(product):
    key = "Product:" + product.id
    return key


def add_object_to_cache(cache, key, product):
    item = CacheItem(product)
    version = cache.add(key, item)
    print("\nObject is added in cache.")
    return version


def update_object_in_cache(cache, key, product, item_version):
    product.units_available = 10
    updated_item = CacheItem(product)
    updated_item.set_cache_item_version(item_version)

    # Update item if version match
    updated_item_version = cache.insert(key, updated_item)
    if item_version.get_version() != updated_item_version.get_version():
        print("\nItem has changed since last time it was fetched.")
    return updated_item_version


def get_object_with_item_version(cache, key, item_version):
    # Get item only if version is superior to old version
    fetched_product = cache.get_if_newer(key, item_version, Product)
    if fetched_product is None:
        print("\nSpecified item version is latest.")
    else:
        print("\nCurrent version of item is: ", item_version.get_version())
        print("\nObject is fetched from cache.")
        print(fetched_product)


def remove_object_with_item_version(cache, key, item_version):
    # Remove item if version match
    removed_product = cache.remove(key, Product, None, None, item_version)
    if removed_product is None:
        print("\nThe item cannot be removed as the newer version of item exists in cache.")
    else:
        print("\nObject is removed from cache.")
        print("\nThe removed object is:")
        print(removed_product)


def dispose_cache(cache):
    cache.close()
    print("\nCache closed.")


# Initialize instance of the cache to perform operations
cache = initialize_cache()

# Create a Product object
product = create_product()
key = get_key(product)

# Adding item
item_version = add_object_to_cache(cache, key, product)

# Update item
latest_item_version = update_object_in_cache(cache, key, product, item_version)

# Get an item with cache item version
get_object_with_item_version(cache, key, item_version)

# Remove item
remove_object_with_item_version(cache, key, latest_item_version)

# Dispose the cache once done
dispose_cache(cache)
